src/components/structures/TimelinePanel.tsx

```diff
@@ -806,16 +806,23 @@ class TimelinePanel extends React.Component<IProps, IState> {
         // Can be null for the notification timeline, etc.
         if (!this.props.timelineSet.room) return;
 
+        if (ev.getRoomId() !== this.props.timelineSet.room.roomId) return;
+
+        if (!this.state.events.includes(ev)) return;
+
+        const firstVisibleEventIndex = this.checkForPreJoinUISI(this.state.events);
+        if (firstVisibleEventIndex !== this.state.firstVisibleEventIndex) {
+            this.setState({ firstVisibleEventIndex });
+        }
```

Comment:
this might get expensive for every decrypt, we might want to throttle this

Reply:
Just that part, otherwise you might miss "valid" events given you only run this section conditionally